### ISSUE TITLE
Pass max_count to troupe decorator dynamically

### ISSUE THREAD

**Issue by andatt** [kquick/Thespian#29]
I am struggling to be able to pass a max_count value to the troupe decorator dynamically. I am calling like this:

```
from thespian.troupe import troupe
from thespain.actors import *

count = 30
decorated_actor= troupe(max_count=count)(Actor)
```
But when I pass the decorated_actor to the actor system it behaves as if its not decorated. What am I doing wrong here?

**Comment by kquick** [comment]
Hi @andatt,

It's a little hard to say without knowing the rest of your example and what the output is, but if you are decorating the standard `Actor`, the default `receiveMessage` will simply discard messages.  Can you share a little more of your implementation and describe what you are seeing and what you expect to see?

Thanks,
  Kevin

**Comment by andatt** [comment]
Hi Kevin

Sorry about the delay in reply, didn't get notification of your response for some reason.

So I have some code like this:

```
@troupe(max_count=30)
class SomeActor
    def receiveMessage(self, msg, sender):
        time.sleep(3)

actor = actor_system.createActor(SomeActor)
messages = ["some message"] * 30

for message in message:
    actor_system.tell(actor, msg)
```
This works as expected by running through the for loop without blocking. The Troupe actors are created and everything works fine.

However when I do this:
```
class SomeActor
    def receiveMessage(self, msg, sender):
        time.sleep(3)

count = 30
decorated_actor= troupe(max_count=count)(SomeActor) 
actor = actor_system.createActor(decoarted_actor)
messages = ["some message"] * 30

for message in message:
    actor_system.tell(actor, msg)
```
Now the tell call blocks on each message waiting for the sleep to complete. So far as I can tell the troupe decorator is now not working. What mistake did I make here?

The reason I want to do this is to have the count value set at run time by the calling code rather than hard coded like in the first example.

Hope that clarifies, if you need any more info let me know!

Thanks

**Comment by kquick** [comment]
I was able to reproduce your problem with some updates to your code, and also to run it with the dynamic count method with some adjustments.

First the version that reproduces the behavior you saw:

    from thespian.actors import *
    from thespian.troupe import troupe
    import time

    class SomeActor(Actor):
        def receiveMessage(self, msg, sender):
            time.sleep(3)

    if __name__ == "__main__":
        actor_system = ActorSystem("multiprocTCPBase")
        count = 30
        decorated_actor = troupe(count)(SomeActor)
        actor = actor_system.createActor(decorated_actor)
        messages = ["some message"] * 30
        for message in messages:
        actor_system.tell(actor, message)
        actor_system.shutdown()

This version should show the ~90-second runtime, and if you add the `@troupe` decorator to the class definition instead of using `decorated_actor`, you get the troupe behavior with about a 10 second runtime.

However, I can get the 10 second runtime using `decorated_actor` as well if I move the `count = ...` and `decorated_actor = ...` lines out of the `if __name__ == "__main__":` body and to the top level:


    from thespian.actors import *
    from thespian.troupe import troupe
    import time

    class SomeActor(Actor):
        def receiveMessage(self, msg, sender):
            time.sleep(3)

    count = 30
    decorated_actor = troupe(count)(SomeActor)

    if __name__ == "__main__":
        actor_system = ActorSystem("multiprocTCPBase")
        actor = actor_system.createActor(decorated_actor)
        messages = ["some message"] * 30
        for message in messages:
        actor_system.tell(actor, message)
        actor_system.shutdown()

The reason this works, but having it protected by the `__main__` test does not is because there is a new process spawned for each Actor created, and the spawned process will re-import the main file, but `__name__` will not be `__main__`.   Thus, the `@troupe` decorator and having `decorated_actor` as top-level globals in the file are equivalent, but placing the `decorate_actor` initialization inside a top-level expression will prevent it from being executed when the subprocess is created and therefore the decoration does not occur.

Thank you for the example, although I did have to make some additional assumptions about your complete configuration, but hopefully these assumptions match your situation and can explain the behavioral differences you are seeing.

**Comment by andatt** [comment]
Thanks very much! That makes sense. My actual code is a little different but I am pretty sure that what you describe is the issue in my case too as I notice the first two actors spawn at a speed you would expect given a troupe and then slow after that.

Given that behaviour it seems like that setting the max_count dynamically can be done but is somewhat restricted in that you can't pass parameters to any max_count calculation (as it needs to be a top level global). I was attempting to load that value from a yaml config file but don't think that will be possible as I need to pass a param in to specify which config item is required. I will need to have a think about that to see if there are any other alternatives.

If the code were to specify some arbitrary high value on max_count (like say 50) across all troupe decorated actors will Thespian manage this without issue or is that likely to cause problems?

**Comment by kquick** [comment]
The limit is mostly that imposed by your OS, not Thespian.  There are various OS limits like the number of processes, number of open file descriptors, etc. that will probably be what you will encounter.  I do not think 50 is likely to be an issue.

In addition, the troupe decorator will only create actors on-demand: if all existing actors are handling work and a new work message arrives, a new actor will be created (subject to the limit), but if there is an idle actor the work will be given to that actor instead.  There is also an idle limit count for the troupe decorator, so if you do something like:

    @troupe(max_count=50, idle_count=3)
    class MyActor(Actor): ...

Then normally there will only be 4 MyActor instances running (the leader plus the 3 workers) but it will create up to 50 workers as needed.

**Comment by andatt** [comment]
I see, that's very helpful. Will close now as my original question has been answered. Thanks again for your help!

**Comment by kquick** [comment]
Hello @andatt,

I don't know if you are still utilizing Thespian but I wanted to let you know that I've just released version 3.10.0 which provides the ability to dynamically adjust the number of troupe workers.

**Comment by andatt** [comment]
Hi @kquick 

Yes I am still using Thespian. That's brilliant news! Thanks Kevin. Will be very handy to be able to do that. Will take a look at the latest update and give it a try.
